src/Test/Utilities/Shared/Syntax/SourceUtilities.cs

```diff
@@ -0,0 +1,90 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+using Microsoft.CodeAnalysis.Text;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace Roslyn.Test.Utilities.Syntax
+{
+    internal sealed class SegmentedSourceText: SourceText
+    {
+        const int BLOCK_SIZE = 1024;
+
+        List<char[]> blocks = new List<char[]>();
+
+        public SegmentedSourceText(IEnumerable<char> from)
+        {
+            blocks.Add(new char[BLOCK_SIZE]);
+            var i = 0;
+
+            foreach (char ch in from)
+            {
+                if (i == BLOCK_SIZE)
+                {
+                    i = 0;
+                    blocks.Add(new char[BLOCK_SIZE]);
+                }
+                blocks.Last()[i] = ch;
+                i++;
+            }
+        }
+
+        public override char this[int position]
+        {
+            get
+            {
+                var block = position / BLOCK_SIZE;
+                var offset = position % BLOCK_SIZE;
+                return blocks[block][offset];
+            }
+        }
+
+        public override Encoding Encoding => Encoding.UTF8;
+
+        public override int Length
+        {
+            get
+            {
+                if (blocks.Count != 0)
+                {
+                    return (blocks.Count - 1) * BLOCK_SIZE + blocks.Last().Length;
+                }
+                else
+                {
+                    return 0;
+                }
+            }
+        }
+
+        public override void CopyTo(int sourceIndex, char[] destination, int destinationIndex, int count)
+        {
+            for (var i = 0; i < count; i++)
+            {
+                destination[destinationIndex + i] = this[sourceIndex + i];
+            }
+        }
+    }
+
+    public static class SourceUtilities
+    {
+        private static IEnumerable<char> RandomString()
+        {
+
+            // Apparently this fixed seed exposed a bug at some point
+            var random = new Random(12345);
+            // 40 million "character"s
+            const int n = 40 * 1000 * 1000;
+
+            for(int i = 0; i < n; i++)
+            {
+                yield return (char)random.Next(char.MaxValue);
+            }
+        }
+
+        public static SourceText GenerateLargeRandomString()
+        {
+            return new SegmentedSourceText(RandomString());
```

Comment:
It looks like SegmentedSourceText keeps all the characters in memory, can this be avoided? For example, if the tests that are using this SourceText never access characters in random order, but always do this consecutively, we could have implementation of SourceText that doesn't cache any characters in memory, but rather "generates" them on demand.